feat: add openapi-ts swr plugin

Pewww opened the pull request:
# SWR v2 Plugin PR Summary

## Overview
New SWR plugin for `@hey-api/openapi-ts` that generates type-safe SWR hooks and utilities from OpenAPI specifications. This plugin follows SWR's official best practices and recommended patterns.

## Key Features

### 1. Core Functionality
- **SWR Keys**: Generates key functions for cache management following SWR's array key pattern
- **useSWR Options**: Creates options helpers for standard queries with automatic key and fetcher generation
- **useSWRMutation Options**: Generates mutation options for POST/PUT/DELETE operations
- **useSWRInfinite Options**: Provides pagination support with automatic page key generation

### 2. Architecture

#### Plugin Structure
```
packages/openapi-ts/src/plugins/swr/
├── config.ts           # Plugin configuration and defaults
├── index.ts            # Main exports
├── plugin.ts           # Plugin handler entry point
├── types.d.ts          # TypeScript type definitions
├── useType.ts          # Type utilities for operation data/response
└── v2/
    ├── plugin.ts              # v2 handler implementation
    ├── swrKey.ts              # Key generation logic
    ├── swrOptions.ts          # useSWR options generation
    ├── swrMutationOptions.ts  # useSWRMutation options generation
    └── swrInfiniteOptions.ts  # useSWRInfinite options generation
```

#### Key Design Decisions

**SWR Key Pattern (v2)**
- Uses array keys with object serialization (SWR 1.1.0+)
- No params: `['/api/users']`
- With params: `['/api/users/{id}', options]`
- Keys always return valid arrays (never null)
- Conditional fetching controlled by consumers

**Query Options (useSWR)**
- Generates `key` property using registered SWR key function
- Generates `fetcher` property with async SDK function call
- Automatically adds `throwOnError: true` to SDK calls
- Handles both required and optional parameters correctly
- Respects SDK plugin's `responseStyle` config (data vs object)

**Mutation Options (useSWRMutation)**
- Follows SWR mutation patterns with key + arg-based fetcher
- Keys: `['/api/users']` or `['/api/users/{id}', options]`
- Fetcher signature: `async (_key, { arg }) => ...`
- Spreads both options and arg into SDK function call

**Infinite Options (useSWRInfinite)**
- Only generated for operations with pagination support
- `getKey` function signature matches SWR requirements
- Merges pagination parameter into query params
- Key structure: `[path, { ...options, query: { ...query, page: pageIndex } }]`
- Fetcher extracts options from key array

### 3. Configuration Options

```typescript
{
  name: 'swr',
  case: 'camelCase',
  comments: true,
  exportFromIndex: false,

  swrKeys: {
    enabled: true,
    exported: true,
    name: '{{name}}Key',
    case: 'camelCase'
  },

  swrOptions: {
    enabled: true,
    exported: true,
    name: '{{name}}Options',
    case: 'camelCase'
  },

  swrMutationOptions: {
    enabled: true,
    exported: true,
    name: '{{name}}Mutation',
    case: 'camelCase'
  },

  swrInfiniteOptions: {
    enabled: true,
    exported: true,
    name: '{{name}}Infinite',
    case: 'camelCase'
  }
}
```

### 4. Generated Code Examples

**Simple Query (No Params)**
```typescript
export const getUsersKey = () => ["/users"];

export const getUsersOptions = () => ({
  key: getUsersKey(),
  fetcher: async () => {
    const { data } = await getUsers({ throwOnError: true });
    return data;
  }
});
```

**Query with Required Params**
```typescript
export const getUserByIdKey = (options: GetUserByIdData) =>
  ["/users/{id}", options];

export const getUserByIdOptions = (options: GetUserByIdData) => ({
  key: getUserByIdKey(options),
  fetcher: async () => {
    const { data } = await getUserById({ ...options, throwOnError: true });
    return data;
  }
});
```

**Mutation**
```typescript
export const createUserMutation = (options?: CreateUserData) => ({
  key: ['/users'],
  fetcher: async (_key, { arg }) => {
    const { data } = await createUser({ ...options, ...arg, throwOnError: true });
    return data;
  }
});
```

**Infinite Query with Pagination**
```typescript
export const getUsersInfinite = (options?: GetUsersData) => ({
  getKey: (pageIndex: number, previousPageData: GetUsersResponse | null) =>
    ['/users', { ...options, query: { ...options?.query, page: pageIndex } }],
  fetcher: async (key: readonly [string, GetUsersData]) => {
    const { data } = await getUsers({ ...key[1], throwOnError: true });
    return data;
  }
});
```

## Implementation Details

### Dependencies
- `@hey-api/sdk` - Required for SDK function generation
- `@hey-api/typescript` - Required for TypeScript type generation
- `axios` (external) - For AxiosError type import

### Special Handling
- **SSE Operations**: Skipped (not supported by SWR)
- **asClass Mode**: Properly handles class-based SDK clients
- **Optional Parameters**: Uses TypeScript optional chaining for safety
- **ESLint Comments**: Adds file-level `eslint-disable` for unused vars to avoid formatter issues

## Testing

### Test Coverage
- Test files located in: `packages/openapi-ts-tests/main/test/plugins/swr/`
- Covers OpenAPI versions: 2.0.x, 3.0.x, 3.1.x
- Tests different client modes: fetch, axios, asClass
- Edge cases tested: no-params, body-path, nullable, path-only

### Snapshot Tests
- Generated code snapshots maintained in test directories
- Recently synced and updated to match latest implementation

## Usage Example

```typescript
// openapi-ts.config.ts
export default {
  client: '@hey-api/client-fetch',
  input: './openapi.json',
  output: './src/client',
  plugins: [
    '@hey-api/sdk',
    '@hey-api/typescript',
    {
      name: 'swr',
      swrOptions: true,
      swrMutationOptions: true,
      swrInfiniteOptions: true,
    }
  ]
};

// In your React component
import useSWR from 'swr';
import useSWRMutation from 'swr/mutation';
import useSWRInfinite from 'swr/infinite';
import { getUsersOptions, createUserMutation, getUsersInfinite } from './client/swr.gen';

// Query
const { data, error } = useSWR(...Object.values(getUsersOptions()));

// Mutation
const { trigger } = useSWRMutation(...Object.values(createUserMutation()));

// Infinite
const { getKey, fetcher } = getUsersInfinite({ query: { status: 'active' } });
const { data, size, setSize } = useSWRInfinite(getKey, fetcher);
```

## Documentation Status
- Documentation page exists at: `docs/openapi-ts/plugins/swr.md`
- Currently marked as "soon" with feature status placeholder
- Needs to be updated with comprehensive usage guide

## PR Checklist

### Code Changes
- [x] Plugin implementation complete
- [x] Type definitions added
- [x] Configuration system implemented
- [x] Test coverage added
- [x] Snapshots updated and synced
- [x] Lint issues resolved

### Testing
- [x] Unit tests for key generation
- [x] Snapshot tests for all OpenAPI versions
- [x] Edge case coverage

## Breaking Changes
None - This is a new plugin, no breaking changes expected.

## Migration Notes
If users were using an unofficial or custom SWR integration, they should:
1. Remove custom SWR helper code
2. Add the official `swr` plugin to their config
3. Update imports to use generated code from `swr.gen.ts`
4. Follow new key patterns (array keys with object serialization)

## Related Issues
- Original feature request: Issue #1479 (referenced in docs)
- Any other related issues should be linked in the PR description

---

Please review the PR and let me know if anything needs to be revised.
I’ve confirmed that the basic queries and mutations work well, but I don’t think the current spec is perfect yet.
Since there are naturally areas to improve, if this PR is merged, I’d like to cautiously suggest releasing it without adding it to the [docs](https://heyapi.dev/openapi-ts/plugins/swr) for now.

Thank you for the great library!

Pewww committed:
feat: add openapi-ts swr plugin

nemo-wrtn committed:
fix: lint

mrlubos commented:
@Pewww `pnpm test:update`

mrlubos commented:
@Pewww the only other question I have is why and how you decided to use that query key design!

Pewww committed:
fix: update snapshot tests

Pewww committed:
fix: re-design swr plugin structure

Pewww committed:
fix: re-design swr plugin structure

Pewww committed:
refactor: handling logic for SWR query key options

Pewww committed:
refactor: handling logic for SWR query key options

Pewww committed:
fix: sync test and snapshots

Pewww committed:
fix: sync test and snapshots

Pewww commented:
@mrlubos Sorry it took a bit of time to finish the revisions.
I’ve applied the changes you mentioned (moving the plugin back into the `v2` folder and switching from `tsc` to the `ts-dsl` syntax).
Could you review current version of code?

Regarding the current query key design: as far as I understand, SWR allows any serializable value as a key.
Internally, it uses logic similar to `unstable_serialize` to stringify arrays and objects into a JSON-based cache key, so the valid range is fairly broad.
We just need to avoid non-pure values like `Date`, `Map`, etc.

Also, when injecting values from `options`, if a key contains dashes, you would need bracket notation to access it.
Rather than handling that inside the plugin, I thought it would be more flexible to just pass the `options` object through as-is.

Separately, it looks like some tests are failing, but they don’t seem related to my changes.
Could we take a look together?
I’m not sure if I need to run an update command on my side.
On my branch, running `test:update`, `examples:generate`, `format`, and `lint:fix` all completed without issues.

---

mrlubos commented:
@Pewww Did you rebuild the package before running those commands?

Pewww committed:
fix: rebuild tests

Pewww committed:
fix: rebuild tests

Pewww commented:
> @Pewww Did you rebuild the package before running those commands?

@mrlubos Oh, I missed it. Thanks.

I’ve been thinking about it, handling `previousPageData` inside the `getKey` function in the `useSWRInfinite` pattern feels tricky.
Depending on what data `previousPageData` holds, we need to decide whether to return `null`, but doing that check directly there seems difficult — both type-wise and functionally.

In contrast, libraries like React Query seems to have a structure that makes it much easier for the plugin to handle such cases.
Do you have any good ideas?
Otherwise, I’m considering simply excluding infinite support for now.

Pewww committed:
fix: remove unusing parameters in swr plugin

Pewww committed:
fix: remove unusing parameters in swr plugin

Pewww committed:
fix: merge conflict

Pewww commented:
@mrlubos Hello, hope you have a great weekend.
I've resolved the type and build errors.
Could you please review it when you get a chance?

mrlubos committed:
refactor: clean up config

mrlubos committed:
refactor: clean up config

Pewww commented:
> nope! Why did you remove the `useSwr()` hooks? Will people not want that?

@mrlubos When I worked on this, I checked how react-query plugin does it, and it provides options like `queryKey` and `queryFn` in a similar manner.

I thought it would be good for Swr to follow a similar approach, and that instead of using `useSwr` directly, exposing options individually and then combining that information would actually be more user-friendly. It also feels like a more flexible design.

If there’s a future need for hooks specifically, I think we can add support at that point.
Please let me know if you see it differently.

renovate[bot] committed:
fix(deps): update dependency vue3-select-component to v0.14.0

github-actions[bot] committed:
ci: release

mrlubos committed:
chore: update lock file

mrlubos committed:
Merge pull request #3059 from hey-api/changeset-release/main

mrlubos committed:
Merge branch 'main' of https://github.com/hey-api/openapi-ts into Pewww/main

mrlubos committed:
refactor: clean up plugin node methods

mrlubos committed:
Merge pull request #3060 from hey-api/refactor/plugin-node

refactor: clean up plugin node methods

JanST123 committed:
Fix: Having paramsSerializer NOT defined should take the query params!

JanST123 committed:
revert version for pull request

mrlubos committed:
chore(deps): upgrade pnpm to 10.25.0

mrlubos committed:
Merge pull request #3063 from hey-api/chore/pnpm-bump-3

chore(deps): upgrade pnpm to 10.25.0

mrlubos committed:
chore: support thunks in doc methods

mrlubos committed:
Merge pull request #3065 from hey-api/chore/doc-thunk

chore: support thunks in doc methods

mrlubos committed:
chore: add lazy dsl node

mrlubos committed:
Merge pull request #3067 from hey-api/chore/dsl-lazy

chore: add lazy dsl node

mrlubos committed:
chore: pass ast context

mrlubos committed:
Merge pull request #3068 from hey-api/chore/ast-context

chore: pass ast context

mrlubos committed:
refactor: move factories into a separate module

mrlubos committed:
Merge pull request #3070 from hey-api/refactor/dsl-factories

refactor: move factories into a separate module

mrlubos committed:
Merge branch 'main' into renovate/vue3-select-component-0.x

renovate[bot] committed:
chore(deps): update dependency ky to v1.14.1

renovate[bot] committed:
chore(deps): update actions/create-github-app-token action to v2.2.1

renovate[bot] committed:
chore(deps): update dependency @changesets/parse to v0.4.2

mrlubos committed:
Merge pull request #2828 from hey-api/renovate/vue3-select-component-0.x

fix(deps): update dependency vue3-select-component to v0.14.0

mrlubos committed:
Merge pull request #3071 from hey-api/renovate/ky-1.x

chore(deps): update dependency ky to v1.14.1

renovate[bot] committed:
chore(deps): update dependency @types/bun to v1.3.4

renovate[bot] committed:
chore(deps): update dependency swr to v2.3.7

renovate[bot] committed:
chore(deps): update dependency @pinia/colada to v0.18.0

renovate[bot] committed:
chore(deps): update dependency fastify to v5.6.2

renovate[bot] committed:
chore(deps): update dependency valibot to v1.2.0

mrlubos committed:
Merge pull request #3074 from hey-api/renovate/actions-create-github-app-token-2.x

chore(deps): update actions/create-github-app-token action to v2.2.1

mrlubos committed:
Merge pull request #3082 from hey-api/renovate/changesets-parse-0.x

chore(deps): update dependency @changesets/parse to v0.4.2

mrlubos committed:
Merge pull request #3081 from hey-api/renovate/valibot-1.x

chore(deps): update dependency valibot to v1.2.0

renovate[bot] committed:
chore(deps): update angular monorepo

renovate[bot] committed:
chore(deps): update dependency @changesets/get-github-info to v0.7.0

mrlubos committed:
Merge pull request #3083 from hey-api/renovate/bun-1.x

chore(deps): update dependency @types/bun to v1.3.4

mrlubos committed:
Merge pull request #3072 from hey-api/renovate/swr-2.x

chore(deps): update dependency swr to v2.3.7

mrlubos committed:
Merge pull request #3076 from hey-api/renovate/pinia-colada-0.x

chore(deps): update dependency @pinia/colada to v0.18.0

mrlubos committed:
Merge pull request #3077 from hey-api/renovate/fastify-5.x

chore(deps): update dependency fastify to v5.6.2

renovate[bot] committed:
chore(deps): update dependency globals to v16.5.0

renovate[bot] committed:
chore(deps): update dependency ofetch to v1.5.1

mrlubos committed:
Merge pull request #3049 from hey-api/renovate/angular-monorepo

chore(deps): update angular monorepo

renovate[bot] committed:
chore(deps): update dependency axios to v1.13.2

mrlubos committed:
Merge pull request #3075 from hey-api/renovate/changesets-get-github-info-0.x

chore(deps): update dependency @changesets/get-github-info to v0.7.0

mrlubos committed:
Merge pull request #3084 from hey-api/renovate/globals-16.x

chore(deps): update dependency globals to v16.5.0

mrlubos committed:
Merge pull request #3085 from hey-api/renovate/ofetch-1.x

chore(deps): update dependency ofetch to v1.5.1

renovate[bot] committed:
chore(deps): update dependency @nuxt/test-utils to v3.21.0

mrlubos committed:
Merge pull request #3089 from hey-api/renovate/axios-1.x

chore(deps): update dependency axios to v1.13.2

renovate[bot] committed:
chore(deps): update dependency tsdown to v0.17.2

mrlubos committed:
Merge pull request #3092 from hey-api/renovate/nuxt-test-utils-3.x

chore(deps): update dependency @nuxt/test-utils to v3.21.0

renovate[bot] committed:
fix(deps): update dependency semver to v7.7.3

renovate[bot] committed:
chore(deps): update dependency eslint-plugin-vue to v9.33.0

mrlubos committed:
Merge pull request #3086 from hey-api/renovate/tsdown-0.x

chore(deps): update dependency tsdown to v0.17.2

renovate[bot] committed:
chore(deps): update dependency yaml to v2.8.2

mrlubos committed:
Merge pull request #3073 from hey-api/renovate/semver-7.x

fix(deps): update dependency semver to v7.7.3

mrlubos committed:
Merge pull request #3094 from hey-api/renovate/eslint-plugin-vue-9.x

chore(deps): update dependency eslint-plugin-vue to v9.33.0

mrlubos committed:
Merge pull request #3088 from hey-api/renovate/yaml-2.x

chore(deps): update dependency yaml to v2.8.2

renovate[bot] committed:
chore(deps): update dependency vue to v3.5.25

renovate[bot] committed:
chore(deps): update actions/checkout action to v6

mrlubos committed:
Merge pull request #3087 from hey-api/renovate/vue-monorepo

chore(deps): update dependency vue to v3.5.25

renovate[bot] committed:
chore(deps): update dependency rollup to v4.53.3

mrlubos committed:
Merge pull request #3098 from hey-api/renovate/actions-checkout-6.x

chore(deps): update actions/checkout action to v6

mrlubos committed:
Merge pull request #3079 from hey-api/renovate/rollup-4.x

chore(deps): update dependency rollup to v4.53.3

renovate[bot] committed:
chore(deps): update eslint monorepo to v9.39.1

renovate[bot] committed:
chore(deps): update vitest monorepo to v3.2.4

mrlubos committed:
Merge pull request #3096 from hey-api/renovate/vitest-monorepo

chore(deps): update vitest monorepo to v3.2.4

mrlubos committed:
Merge pull request #3095 from hey-api/renovate/eslint-monorepo

chore(deps): update eslint monorepo to v9.39.1

mrlubos committed:
feat(sdk): lazily initialize sub-resources to improve performance

mrlubos committed:
Merge pull request #3099 from hey-api/feat/sdk-instance-lazy

feat(sdk): lazily initialize sub-resources to improve performance

mrlubos committed:
refactor: clean up class name logic

mrlubos committed:
Merge pull request #3101 from hey-api/refactor/sdk-class-name

refactor: clean up class name logic

mrlubos committed:
fix(output): sanitize reserved names with underscore suffix instead of prefix

mrlubos committed:
Merge pull request #3102 from hey-api/refactor/sdk-operation-name

fix(output): sanitize reserved names with underscore suffix instead o…

Commit:
Update package.json

mrlubos committed:
Merge pull request #3103 from hey-api/mrlubos-patch-1

mrlubos committed:
refactor(sdk): clean up instance type so there's no need to cast

mrlubos committed:
Merge pull request #3104 from hey-api/refactor/sdk-class-instance

refactor(sdk): clean up instance type so there's no need to cast

renovate[bot] committed:
chore(deps): update dependency @pinia/colada to v0.18.1

renovate[bot] committed:
chore(deps): update dependency tsdown to v0.17.3

mrlubos committed:
Merge pull request #3106 from hey-api/renovate/pinia-colada-0.x

chore(deps): update dependency @pinia/colada to v0.18.1

mrlubos committed:
Merge pull request #3107 from hey-api/renovate/tsdown-0.x

chore(deps): update dependency tsdown to v0.17.3

JanST123 committed:
Add tests, proving that  params are correctly passed to axios as

JanST123 committed:
update snapshots after fixing axios  ->  mapping

JanST123 committed:
update examples after fixing axios  ->  mapping

mrlubos committed:
Modify axios client to utilize `query` option

Update axios client behavior to use the `query` option when no `paramsSerializer` is provided.

mrlubos committed:
Merge pull request #3062 from JanST123GS/bugfix/axios_query

Fix: In axios client, proving NO paramsSerializer should use the `query` option for params

mrlubos committed:
fix(@hey-api/sdk): dedup classes

github-actions[bot] committed:
chore(docs): update contributors list

mrlubos committed:
chore: port asClass functionality

mrlubos committed:
chore: port function functionality

mrlubos committed:
chore: port operation classes function

Copilot committed:
Initial plan

Copilot committed:
Add support for moduleResolution node16 to behave like nodenext

Co-authored-by: mrlubos <12529395+mrlubos@users.noreply.github.com>

Copilot committed:
Initial plan

mrlubos committed:
refactor: clean up ast context

mrlubos committed:
docs: normalize changesets

Copilot committed:
Add test snapshots for node16 moduleResolution

Co-authored-by: mrlubos <12529395+mrlubos@users.noreply.github.com>

mrlubos committed:
Merge pull request #3118 from hey-api/docs/changesets-subject

docs: normalize changesets

Copilot committed:
Add test case for nested allOf discriminator bug

Co-authored-by: mrlubos <12529395+mrlubos@users.noreply.github.com>

Copilot committed:
Fix discriminator resolution for nested allOf compositions

Add helper function to recursively find discriminators in allOf schemas and update parseAllOf to use it for both OpenAPI 3.0.x and 3.1.x

Co-authored-by: mrlubos <12529395+mrlubos@users.noreply.github.com>

Copilot committed:
Add logic to skip discriminator values for intermediate schemas in multi-level inheritance

Prevent adding discriminator values to schemas that are extended by other schemas in the same discriminator mapping, fixing the never type issue in nested allOf hierarchies

Co-authored-by: mrlubos <12529395+mrlubos@users.noreply.github.com>

Copilot committed:
Add snapshot tests for nested allOf discriminator fix

Added test scenarios for discriminator-allof-nested.json in both 3.0.x and 3.1.x test suites with generated snapshots validating the fix

Co-authored-by: mrlubos <12529395+mrlubos@users.noreply.github.com>

mrlubos committed:
Create thirty-shoes-judge.md

mrlubos committed:
Merge pull request #3115 from hey-api/copilot/fix-module-resolution-node16

Fix: Support moduleResolution "node16" alongside "nodenext"

Copilot committed:
Fix discriminator union types for intermediate schemas

Changed logic to include all child discriminator values in a union for intermediate schemas (e.g., CarDto now has $type: 'Car' | 'Volvo' instead of no discriminator). Removed unused helper function.

Co-authored-by: mrlubos <12529395+mrlubos@users.noreply.github.com>

Copilot committed:
Initial plan

Copilot committed:
feat: Add extension fields (x-*) support to IR types and parsers

Co-authored-by: mrlubos <12529395+mrlubos@users.noreply.github.com>

Copilot committed:
Simplify discriminator logic to use schema's own value only

Removed getAllDiscriminatorValues and union logic. Now each schema gets only its own discriminator value from the mapping, matching the requested output format.

Co-authored-by: mrlubos <12529395+mrlubos@users.noreply.github.com>

Copilot committed:
fix: Correct field ordering in spec types for linter

Co-authored-by: mrlubos <12529395+mrlubos@users.noreply.github.com>

Copilot committed:
Changes before error encountered

Co-authored-by: mrlubos <12529395+mrlubos@users.noreply.github.com>

Copilot committed:
refactor: Reuse parseExtensions and remove JSDoc from extension fields

- Made parseExtensions generic and exportable from schema parsers
- Reused parseExtensions in parameter and operation parsers
- Removed JSDoc comments from extension field index signatures

Co-authored-by: mrlubos <12529395+mrlubos@users.noreply.github.com>

mrlubos committed:
Create proud-berries-matter.md

Copilot committed:
Fix TypeScript errors in discriminator implementation

- Changed DiscriminatorInfo.values type from string[] to ReadonlyArray<string> to match discriminatorValues return type
- Added null check for array item access to prevent possibly undefined errors
- Changed required array mutation from push to spread operator to work with readonly arrays

Co-authored-by: mrlubos <12529395+mrlubos@users.noreply.github.com>

Copilot committed:
refactor: Extract SpecificationExtensions type and apply to all spec interfaces

- Created SpecificationExtensions interface in 3.0.x and 3.1.x spec.d.ts
- Applied to all 27 interfaces that support extensions in each version
- Replaced inline `[extension: x-${string}]: unknown` with extends clause
- Also added SpecificationExtensions to JsonSchemaDraft2020_12

Co-authored-by: mrlubos <12529395+mrlubos@users.noreply.github.com>

Copilot committed:
fix: Add missing parseExtensions imports in 3.1.x parsers

Co-authored-by: mrlubos <12529395+mrlubos@users.noreply.github.com>

Copilot committed:
Use non-null assertions in for loops and Array/ReadonlyArray syntax

- Replaced unnecessary null checks with non-null assertions (!) in for loops accessing array items
- Changed all [] type syntax to Array<T> or ReadonlyArray<T> format for consistency

Co-authored-by: mrlubos <12529395+mrlubos@users.noreply.github.com>

Copilot committed:
fix: Resolve typecheck errors with extension field types

- Changed parseExtensions to use any types instead of generic constraints
- Removed SpecificationExtensions from interfaces with index signatures (CallbackObject, PathsObject, ResponsesObject)
- Changed PathsObject index signature from /${string} to string for compatibility
- These changes fix TypeScript errors while preserving extension field support

Co-authored-by: mrlubos <12529395+mrlubos@users.noreply.github.com>

mrlubos committed:
Merge pull request #3117 from hey-api/copilot/fix-discriminator-in-allof

Fix discriminator resolution in nested allOf inheritance hierarchies

Copilot committed:
fix: Correct extension field type handling per review feedback

- Changed parseExtensions to use Record<string, unknown> instead of any
- Added type assertions at call sites for type safety
- Restored SpecificationExtensions to CallbackObject, PathsObject, ResponsesObject
- Kept PathsObject with [path: /${string}] template literal type
- Extended index signatures with | unknown to satisfy TypeScript

Co-authored-by: mrlubos <12529395+mrlubos@users.noreply.github.com>

mrlubos committed:
chore: fix types

mrlubos committed:
Merge pull request #3119 from hey-api/copilot/add-extension-fields-support

feat: Propagate OpenAPI extension fields (x-*) to IR and plugins

github-actions[bot] committed:
ci: release

mrlubos committed:
Merge pull request #3100 from hey-api/changeset-release/main

ci: release

Copilot committed:
Initial plan

Copilot committed:
Fix discriminator inheritance to use Omit for proper type narrowing

- Remove union of child discriminator values from intermediate types
- Each type now gets only its direct discriminator value
- Use Omit<> to remove parent discriminator properties and avoid conflicts
- Enables proper TypeScript type narrowing with discriminated unions

Example: CarDto.$type is now 'Car' instead of 'Car' | 'Volvo'
VolvoDto uses Omit<CarDto, '$type'> & { $type: 'Volvo' }

Co-authored-by: mrlubos <12529395+mrlubos@users.noreply.github.com>

mrlubos committed:
fix: import styles for vue3-select-component

mrlubos committed:
Merge pull request #3121 from hey-api/fix/vue3-select-component-styles

fix: import styles for vue3-select-component

Copilot committed:
Update test snapshots and regenerate examples for Omit discriminator changes

- Updated discriminator test snapshots (3.0.x and 3.1.x) for:
  - discriminator-all-of
  - discriminator-allof-nested
  - transformers-all-of
- Regenerated openapi-ts-fetch example with updated types
- All snapshots now reflect the Omit<> wrapper for discriminator properties

Co-authored-by: mrlubos <12529395+mrlubos@users.noreply.github.com>

Copilot committed:
Fix baseUrl in test snapshots

- Updated baseUrl template in snapshots to match generated output
- Changed from ${string}://specs to ${string}://${string}
- All discriminator tests now pass

Co-authored-by: mrlubos <12529395+mrlubos@users.noreply.github.com>

mrlubos committed:
Enhance type narrowing in discriminated types

Improved type narrowing in discriminated types for better type safety.

mrlubos committed:
Merge pull request #3120 from hey-api/copilot/fix-wrong-discriminator-issue

bilby91 committed:
fix: use serialized body in SSE client requests

SSE client calls were using raw `opts.body` instead of properly serialized body data, causing SSE requests with JSON bodies to send malformed data. Now correctly passes `serializedBody` (via `getValidRequestBody`) to `createSseClient` across all affected client adapters (fetch, angular, axios, ky, next, nuxt).

mrlubos committed:
docs: clean up logos

mrlubos committed:
Merge pull request #3124 from hey-api/docs/github-logos

docs: clean up logos

Commit:
Revert "Fix: In axios client, proving NO paramsSerializer should use the `query` option for params"

mrlubos committed:
Create khaki-experts-confess.md

mrlubos committed:
test: update snapshots

mrlubos committed:
Merge pull request #3125 from hey-api/revert-3062-bugfix/axios_query

Revert "Fix: In axios client, proving NO paramsSerializer should use the `query` option for params"

github-actions[bot] committed:
ci: release

mrlubos committed:
Merge pull request #3122 from hey-api/changeset-release/main

ci: release

mrlubos committed:
test: update snapshots

mrlubos committed:
chore: update Angular plugin

renoschubert committed:
fix(client-nuxt): preserve null in WithRefs type for nullable fields

The WithRefs type was using NonNullable which stripped null from
nullable field types. This made it impossible to pass null values
to API endpoints that expect nullable fields (e.g., `street: null | string`).

Added `| Extract<TData[K], null>` to both branches of WithRefs to
preserve null in the union when the original type includes null.

Fixes: nullable fields in request body cannot be set to null

renovate[bot] committed:
chore(deps): update dependency swr to v2.3.8

renovate[bot] committed:
fix(deps): update dependency c12 to v3.3.3

renovate[bot] committed:
chore(deps): update dependency @pinia/colada to v0.19.1

renovate[bot] committed:
chore(deps): update dependency rollup to v4.54.0

renovate[bot] committed:
chore(deps): update dependency turbo to v2.7.0

renovate[bot] committed:
chore(deps): update pnpm to v10.26.1

renovate[bot] committed:
chore(deps): update stefanzweifel/git-auto-commit-action action to v7.1.0

mrlubos committed:
Merge pull request #3136 from hey-api/renovate/swr-2.x

mrlubos committed:
Merge pull request #3139 from hey-api/renovate/rollup-4.x

mrlubos committed:
Merge pull request #3141 from hey-api/renovate/turbo-monorepo

renovate[bot] committed:
chore(deps): update dependency @types/bun to v1.3.5

mrlubos committed:
Merge pull request #3138 from hey-api/renovate/pinia-colada-0.x

renovate[bot] committed:
chore(deps): update dependency arktype to v2.1.29

mrlubos committed:
Merge pull request #3134 from hey-api/renovate/bun-1.x

mrlubos committed:
Merge pull request #3143 from hey-api/renovate/stefanzweifel-git-auto-commit-action-7.x

mrlubos committed:
Merge pull request #3137 from hey-api/renovate/c12-3.x

mrlubos committed:
Merge pull request #3142 from hey-api/renovate/pnpm-10.x

mrlubos committed:
Merge pull request #3135 from hey-api/renovate/arktype-2.x

mrlubos committed:
feat: support number resolvers

mrlubos committed:
Merge pull request #3145 from hey-api/feat/plugin-resolvers-number

feat: support number resolvers

mrlubos committed:
fix: add naive lower and upper bound checks

mrlubos committed:
feat: add number base resolvers

mrlubos committed:
docs: update copy

mrlubos committed:
Merge pull request #3148 from hey-api/docs/refresh

docs: update copy

mrlubos committed:
refactor: clean up valibot resolvers

mrlubos committed:
refactor: clean up symbolOnce interface

mrlubos committed:
refactor: clean up valibot resolver interface

mrlubos committed:
refactor: clean up zod resolvers

mrlubos committed:
feat: add number base resolvers

mrlubos committed:
docs: add Resolvers API

mrlubos committed:
Merge pull request #3147 from hey-api/feat/plugin-resolvers-number-2

feat: add number base resolvers

mrlubos committed:
chore: expose file header

renovate[bot] committed:
chore(deps): update dependency tsdown to v0.18.3

mrlubos committed:
chore: add log utilities

mrlubos committed:
chore: clean up operations config

mrlubos committed:
chore: angular plugin structure

Commit:
test: update client-nuxt snapshots

mrlubos committed:
docs: minor tweak to copy

mrlubos committed:
Merge pull request #3160 from hey-api/docs/copy-minor

docs: minor tweak to copy

mrlubos committed:
Merge branch 'main' into renovate/tsdown-0.x

mrlubos committed:
Update changelog for client-nuxt plugin fix

renovate[bot] committed:
chore(deps): update pnpm to v10.26.2

mrlubos committed:
Merge pull request #3140 from hey-api/renovate/tsdown-0.x

chore(deps): update dependency tsdown to v0.18.3

mrlubos committed:
Merge pull request #3161 from hey-api/renovate/pnpm-10.x

chore(deps): update pnpm to v10.26.2

mrlubos committed:
Merge pull request #3131 from rigtigeEmil/fix/client-nuxt-withrefs-nullable

github-actions[bot] committed:
chore(docs): update contributors list

renovate[bot] committed:
chore(deps): update dependency ky to v1.14.2

renovate[bot] committed:
chore(deps): update dependency turbo to v2.7.2

mrlubos committed:
Merge pull request #3164 from hey-api/renovate/ky-1.x

chore(deps): update dependency ky to v1.14.2

mrlubos committed:
Merge branch 'main' into renovate/turbo-monorepo

mrlubos committed:
Merge pull request #3165 from hey-api/renovate/turbo-monorepo

chore(deps): update dependency turbo to v2.7.2

mrlubos committed:
chore: fallback to path in operation id

mrlubos committed:
fix: allow overwriting symbol nodes

mrlubos committed:
chore: clean up angular

mrlubos committed:
docs: update SDK plugin manual

mrlubos committed:
Merge pull request #3109 from hey-api/fix/sdk-class-conflict

Structure API

renovate[bot] committed:
chore(deps): update dependency vitepress-plugin-llms to v1.10.0

mrlubos committed:
docs: fix formatting on sdk plugin page

mrlubos committed:
Merge pull request #3168 from hey-api/docs/sdk-copy-fix

docs: fix formatting on sdk plugin page

mrlubos committed:
fix: simplify symbol merging logic

mrlubos committed:
Merge pull request #3169 from hey-api/fix/core-symbol-merge

fix: simplify symbol merging logic

mrlubos committed:
chore: update to the latest OpenCode snapshot

mrlubos committed:
Merge pull request #3170 from hey-api/chore/opencode-snapshot

chore: update to the latest OpenCode snapshot

mrlubos committed:
Merge pull request #3123 from crunchloop/fix/sse-body-serialization

fix: use serialized body in SSE client requests

mrlubos committed:
chore: update changesets

mrlubos committed:
Merge pull request #3171 from hey-api/docs/changesets-clients

chore: update changesets

mrlubos committed:
Merge pull request #3163 from hey-api/renovate/vitepress-plugin-llms-1.x

chore(deps): update dependency vitepress-plugin-llms to v1.10.0

github-actions[bot] committed:
ci: release

renovate[bot] committed:
chore(deps): update dependency @opencode-ai/sdk to v1.0.216

renovate[bot] committed:
chore(deps): update dependency tsdown to v0.18.4

mrlubos committed:
chore: update lock file

mrlubos committed:
Merge pull request #3146 from hey-api/changeset-release/main

ci: release

mrlubos committed:
Merge branch 'main' into renovate/tsdown-0.x

mrlubos committed:
Merge branch 'main' into renovate/opencode-ai-sdk-1.x

mrlubos committed:
chore: fix snapshots

mrlubos committed:
Merge pull request #3174 from hey-api/chore/fix-snapshot

chore: fix snapshots

mrlubos committed:
Merge branch 'main' into renovate/tsdown-0.x

mrlubos committed:
Merge branch 'main' into renovate/opencode-ai-sdk-1.x

mrlubos committed:
Merge pull request #3173 from hey-api/renovate/tsdown-0.x

chore(deps): update dependency tsdown to v0.18.4

mrlubos committed:
Merge pull request #3172 from hey-api/renovate/opencode-ai-sdk-1.x

chore(deps): update dependency @opencode-ai/sdk to v1.0.216

github-actions[bot] committed:
chore(docs): update contributors list

mrlubos committed:
feat: SDK examples

mrlubos committed:
Merge pull request #3175 from hey-api/feat/sdk-examples

feat: SDK examples

renovate[bot] committed:
chore(deps): update dependency @opencode-ai/sdk to v1.0.221

renovate[bot] committed:
chore(deps): update pnpm to v10.27.0

mrlubos committed:
Merge pull request #3177 from hey-api/renovate/pnpm-10.x

chore(deps): update pnpm to v10.27.0

mrlubos committed:
Merge pull request #3176 from hey-api/renovate/opencode-ai-sdk-1.x

chore(deps): update dependency @opencode-ai/sdk to v1.0.221

github-actions[bot] committed:
ci: release

mrlubos committed:
chore: update lock file

mrlubos committed:
Merge pull request #3178 from hey-api/changeset-release/main

ci: release

mrlubos committed:
Merge branch 'main' of https://github.com/Pewww/openapi-ts into Pewww/main

mrlubos commented:
oh no 🎢

SukkaW commented:
```ts
// Query
const { data, error } = useSWR(...Object.values(getUsersOptions()));

// Mutation
const { trigger } = useSWRMutation(...Object.values(createUserMutation()));

// Infinite
const { getKey, fetcher } = getUsersInfinite({ query: { status: 'active' } });
const { data, size, setSize } = useSWRInfinite(getKey, fetcher);
```

IMHO, this usage is not very friendly. It would be nicer to code-gen wrapped hooks, something like:

```ts
// Query
const { data, error } = useUserOptions({ ...swrOptions });
const { data, error } = useImmutableUserOptions({ ...swrOptions }); // this should be powered by useSWRImmutable

// Mutation
const { mutate } = useUserOptions({ ...swrOptions });
const { trigger } = useMutateUser({ ...swrMutationOptions });

// Infinite
const { data, size, setSize } = useUserInfinite();
```

Also, what happened with dynamic params?

And, it would be nice to be able to access some low-level SWR APIs like middlewares. E.g., the token may be from a React Context like API (`useAuth()`). So we might wanna inject the token to SWR (via SWR key or SWR middleware?)

mrlubos commented:
I think that's a vibe coded description but rest assured this plugin won't have that kind of API

mrlubos commented:
@Pewww seems @SukkaW is making progress on this plugin so please hold off on opening a pull request if you haven't started yet
